### PR TITLE
Validate checksum for AWS S3 signed URL downloads during localization [BT-257]

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala
Comment (breilly2) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala:63]: Should this have a retry disposition like the cases above?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala
Comment (jgainerdewar) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:88]: Does this mean we could theoretically continue retrying forever, if we get a 408 or 429? Or is there a timeout elsewhere in the system?

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala
Comment (jgainerdewar) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala:63]: I also found this unintuitive, given that these two defined `RetryDisposition` types actually represent cases where we do NOT want to do a "normal" limited-number-of-attempts retry. Maybe make the third one `PersistentRetryDisposition`?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/Downloader.scala:13]: Should this extend `DownloadFailure` instead of `DownloadResult`?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala
Comment (breilly2) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:88]: TOL (thinking out loud): I actually wonder if the notion of a class of error that doesn't count against the retry count is over-complicating things. Sure, someone could get unlucky through a sequence of different types of errors, but that'll probably be a rare case.

429 is an interesting one. I could imagine that we might want a completely different retry strategy for that (since it's specifically related to rate limiting)... in other words, differently configured retry intervals and allowed number of retries. Ultimately, we're trying to codify a heuristic for a sensible level of persistence for this context. I think some day we could consider refactoring the code such that the rules are isolated from the actions, and keep the rules simple in the mean time.

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala
Comment (breilly2) [reply cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsPathResolver.scala]: I agree that being explicit here is good.

TOL: In other contexts, if I were to choose one of these to be the assumed default, it would probably be `FatalRetryDisposition` because I think `RuntimeException` itself carries that meaning pretty well. However, when considering how `maybeRetryForDownloadFailure` should handle exceptions without any disposition (and based on recent experience with problems stemming from needlessly retrying "bad request" type errors), I'm inclined to agree with keeping the default disposition-less exception handling as written in this PR.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala
Comment (jgainerdewar) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:88]: Hard to judge the relative badness of the theoretical possibility we could hang forever vs. the actual cases where we'll fail a download due to transient errors we could have waited out, without knowing how often the latter happens. Since (I think?) this is happening on the VM and not in the main Cromwell process I'm less worried about it, but I also agree with Brian that keeping things simple until we have a "real" strategy we fully believe in would be good. There's also the option of a third, higher attempt limit for transient errors if we want to easily keep behavior similar to what's here while preventing infinite retries, though that's kind of icky. I assume an overall download timeout is out of scope for this particular PR.

I slightly lean toward either pulling out the special transient error handling or adding a third retry counter 🙄 . I don't feel like I have a good enough understanding of user expectations (how annoying is it to manually restart a failed workflow vs have one hang for a long time?) and how often we see these transient errors to have a preference between those. That said, I'm OK approving this as-is if you guys are.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:75]: I wonder if we should reset `downloadAttempt` to `0` here. We know we've successfully downloaded the file once, so we can be pretty confident we can do it again. It would be a shame to get that far and then fail the job an a transient error.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GetmChecksum.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GetmChecksum.scala:14]: Do we also need to sanitize for `\`s?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:30]: Nice 👍

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:88]: @jgainerdewar yes this code runs on the VM and not in the main Cromwell process. Someday there will (may) be analogous code that will run in the Cromwell process per BT-389, but this is not that. 🙂

File: project/Settings.scala
Comment (breilly2) [inline project/Settings.scala:146]: Nice 👍

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfiguration.scala
Comment (breilly2) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfiguration.scala:23]: Sidecar change?

Comment (mcovarr) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfiguration.scala:23]: guilty as charged.

File: cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala:156]: I see this is where the transient errors come from. Where do the regular errors come from?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GetmChecksum.scala
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GetmChecksum.scala:14]: I'm starting to think this might be better as an "allow list" of base64  characters instead.

File: cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala:197]: Nit: unindent
```suggestion
        IO.raiseError(new RuntimeException("testing: fatal error") with FatalRetryDisposition)
```

File: cromwell-drs-localizer/src/test/scala/drs/localizer/downloaders/AccessUrlDownloaderSpec.scala
Comment (breilly2) [inline cromwell-drs-localizer/src/test/scala/drs/localizer/downloaders/AccessUrlDownloaderSpec.scala:26]: Nice job characterizing `getm`s output behavior! This will make it easy to upgrade with confidence in the future!

File: cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala
Comment (mcovarr) [reply cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala:156]: line 165

**Comment by jgainerdewar** [comment]
LGTM, withholding thumb since there's a possibility of additional code changes related to the transient retry logic.

**Review by breilly2 — APPROVED** [review]
👍   Looks good. Nice tests!

Most of my feedback, other than the discussions about how complex we want the retry logic to be initially, is minor and/or questions.

The only thing I'd like to definitely answer before merge is the question of whether we should reset `downloadAttempt` on checksum failure.

**Review by jgainerdewar — APPROVED** [review]
👏 👏

**Pull request merged**
